constexpr all the things in irange.h

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147569

I got complaints while irangeifying some files in ExecuTorch
that irange could not be used in a constexpr function. This made the
complaints go away.

I added a constexpr function in irange_test that used to fail to build
with `error: variable of non-literal type 'iterator' (aka
'integer_iterator<int, true>') cannot be defined in a constexpr
function before C++23` and now builds fine.

Differential Revision: [D69959614](https://our.internmc.facebook.com/intern/diff/D69959614/)

Commit:
constexpr all the things in irange.h

I got complaints while irangeifying some files in ExecuTorch
that irange could not be used in a constexpr function. This made the
complaints go away.

I added a constexpr function in irange_test that used to fail to build
with `error: variable of non-literal type 'iterator' (aka
'integer_iterator<int, true>') cannot be defined in a constexpr
function before C++23` and now builds fine.

Differential Revision: [D69959614](https://our.internmc.facebook.com/intern/diff/D69959614/)

[ghstack-poisoned]

Skylion007 commented on c10/util/irange.h:
```diff
@@ -116,7 +116,7 @@ integer_range<Integer2> irange(Integer1 begin, Integer2 end) {
     typename Integer,
     std::enable_if_t<std::is_integral_v<Integer>, bool> = true>
-integer_range<Integer, true> irange(Integer end) {
+constexpr integer_range<Integer, true> irange(Integer end) {
```
Why isn't the the open range above constexpr too? std::max has been constexpr since C++14?

swolchok replied:
I missed a spot, thanks!

Commit:
Update on "constexpr all the things in irange.h"


I got complaints while irangeifying some files in ExecuTorch
that irange could not be used in a constexpr function. This made the
complaints go away.

I added a constexpr function in irange_test that used to fail to build
with `error: variable of non-literal type 'iterator' (aka
'integer_iterator<int, true>') cannot be defined in a constexpr
function before C++23` and now builds fine.

Differential Revision: [D69959614](https://our.internmc.facebook.com/intern/diff/D69959614/)

[ghstack-poisoned]

swolchok committed:
Update on "constexpr all the things in irange.h"


I got complaints while irangeifying some files in ExecuTorch
that irange could not be used in a constexpr function. This made the
complaints go away.

I added a constexpr function in irange_test that used to fail to build
with `error: variable of non-literal type 'iterator' (aka
'integer_iterator<int, true>') cannot be defined in a constexpr
function before C++23` and now builds fine.

Differential Revision: [D69959614](https://our.internmc.facebook.com/intern/diff/D69959614/)

[ghstack-poisoned]

swolchok commented:
looks like I need to send a new PR because my internal git config was messed up by an ExecuTorch script

swolchok commented:
replaced by #147633
